src/IterTools.jl

```diff
@@ -1031,4 +1028,58 @@ function iterate(fs::FieldValues, state=1)
     return (getfield(fs.x, state), state + 1)
 end
 
+# IMerge
+
+"""
+    imerge(a,b, predicate = <=, fa = identity, fb = identity)
+
+Iterate over the union of `a` and `b`, merge-sort style.
+
+Input:
+ - `predicate(ak,bk) -> Bool`:
+    Whether to pick the next element of `a` (true) or `b` (false).
+ - `fa(ak)`, `fb(bk)`: Functions to apply to the picked elements
+
+```jldoctest
+julia> collect(imerge(1:2:5, 2:2:6, <=, identity, x->-x))
+6-element Vector{Int64}:
+  1
+ -2
+  3
+ -4
+  5
+ -6
+```
+"""
+imerge(a,b, p = <=, fa = identity, fb = identity) = IMerge(a,b,p,fa,fb)
+
+struct IMerge{A,B,P,FA,FB}
+    a::A
+    b::B
+    predicate::P
+    fa::FA
+    fb::FB
+end
+
+Base.IteratorSize(::Type{<:IMerge{A,B}}) where {A,B} = longest(Base.IteratorSize.((A,B))...)
+Base.length(m::IMerge) = length(m.a) + length(m.b)
+Base.IteratorEltype(::Type{<:IMerge}) = Base.EltypeUnknown()
```

Comment:
If we got rid of `fa` and `fb`
(which roughly speaking can be replaced with `Iterators.map` in post), this could be
```suggestion
Base.IteratorEltype(::Type{<:IMerge{A,B}}) where {A, B}= Union{Base.IteratorEltype(A), Base.IteratorEltype(B)}
```